### PR TITLE
Closes #1758: Use Firebase remote config for rate / share app

### LINKED ISSUE #1758: Contextual Hints: Clean up and use Firebase remote config
1. For promotions only. Include the content and time for dialogues and notifications
2. May need to add new timing/criterial for above

### LINKED ISSUE #1759: Firebase Telemetry to include some Rocket specific events

### PR DESCRIPTION
**Pull request by cnevinc**
Closes #1758

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/utils/AppConfigWrapper.java [hunk 1]
```diff
@@ -5,28 +5,30 @@
...
+        return FirebaseHelper.getRcLong(context, FirebaseHelper.RATE_APP_NOTIFICATION_THRESHOLD);
     }
 
-    public static int getShareDialogLaunchTimeThreshold(final boolean needExtend) {
+    public static long getShareDialogLaunchTimeThreshold(Context context, final boolean needExtend) {
         if (needExtend) {
-            return DialogUtils.APP_CREATE_THRESHOLD_FOR_SHARE_DIALOG +
-                    DialogUtils.APP_CREATE_THRESHOLD_FOR_RATE_NOTIFICATION -
-                    DialogUtils.APP_CREATE_THRESHOLD_FOR_RATE_DIALOG;
+            return getRateDialogLaunchTimeThreshold(context) +
```
Comment (mTwTm) [inline app/src/main/java/org/mozilla/focus/utils/AppConfigWrapper.java]: The change in diff calculation does not match?

File: firebase/src/firebase/java/org/mozilla/focus/utils/FirebaseWrapper.java [hunk 2]
```diff
@@ -73,6 +73,8 @@
     public static String getRcString(@NonNull Context context, @NonNull String key) {
         if (instance == null) {
             Log.e(TAG, "getRcString: failed, FirebaseWrapper not initialized");
+            throwRcNotInitException();
+
             return "";
```
Comment (mTwTm) [inline firebase/src/firebase/java/org/mozilla/focus/utils/FirebaseWrapper.java]: Avoid magic number? Maybe FIREBASE_STRING_NOT_FOUND or FIREBASE_STRING_DEFAULT

File: firebase/src/firebase/java/org/mozilla/focus/utils/FirebaseWrapper.java [hunk 3]
```diff
@@ -81,17 +83,47 @@ public static String getRcString(@NonNull Context context, @NonNull String key)
...
         }
+        throwGetValueException("getRcString");
         return "";
     }
 
+    static long getRcLong(Context context, String key) {
+        if (instance == null) {
+            Log.e(TAG, "getRcString: failed, FirebaseWrapper not initialized");
+            throwRcNotInitException();
+            return 0L;
```
Comment (mTwTm) [inline firebase/src/firebase/java/org/mozilla/focus/utils/FirebaseWrapper.java]: Avoid magic number? Maybe FIREBASE_LONG_NOT_FOUND or FIREBASE_LONG_DEFAULT

File: firebase/src/firebase_no_op/java/org/mozilla/focus/utils/FirebaseWrapper.java [hunk 4]
```diff
@@ -23,19 +23,29 @@ static FirebaseWrapper getInstance() {
         return instance;
     }
 
-    // get Remote Config string
+    static long getRcLong(Context context, String key) {
+        if (instance == null) {
+            return 0L;
```
[the remark this answers is not in the document]
Comment (cnevinc) [reply firebase/src/firebase_no_op/java/org/mozilla/focus/utils/FirebaseWrapper.java]: fixed

File: firebase/src/firebase_no_op/java/org/mozilla/focus/utils/FirebaseWrapper.java [hunk 5]
```diff
@@ -23,19 +23,29 @@ static FirebaseWrapper getInstance() {
...
+        } else if (value instanceof Long) {
+            return (Long) value;
+        }
+        return 0L;
+    }
+
     // get Remote Config string
     static String getRcString(Context context, String key) {
         if (instance == null) {
             return "";
```
[the remark this answers is not in the document]
Comment (cnevinc) [reply firebase/src/firebase_no_op/java/org/mozilla/focus/utils/FirebaseWrapper.java]: fixed

File: firebase/src/firebase/java/org/mozilla/focus/utils/FirebaseWrapper.java (hunk 2 above)
Comment (cnevinc) [reply firebase/src/firebase/java/org/mozilla/focus/utils/FirebaseWrapper.java]: fixed

File: firebase/src/firebase/java/org/mozilla/focus/utils/FirebaseWrapper.java (hunk 3 above)
Comment (cnevinc) [reply firebase/src/firebase/java/org/mozilla/focus/utils/FirebaseWrapper.java]: fixed

**Comment by mTwTm** [comment]
I just noticed one critical thing: enableRemoteConfig() may run AFTER runPromotion() so it does not really make sense to do things here. If so, I think we should rethink if we want to implement this in this milestone.

**Comment by cnevinc** [comment]
In that case, the user will still get the default behavior. So it's fine.

**Pull request merged**